src/ddsrt/src/string.c

```diff
@@ -99,7 +99,9 @@ ddsrt_strlcpy(
     if (size <= srclen) {
       len = size - 1;
     }
-    memcpy(dest, src, len);
+    if (len > 0) {
+      memcpy(dest, src, len);
```

Comment:
Here I simply have yet another problem with MISRA. (Nothing new here.)

Per the C spec (I use the draft, just like everyone else), 7.21.1 paragraph 2, concerning string function conventions, of which `memcpy` is one: "Where an argument declared as **size_t n** specifies the length of the array for a function, **n** can have the value zero on a call to that function." And further: "Unless explicitly stated otherwise in the description of a particular function in this subclause, pointer arguments on such a call shall still have valid values, as described in 7.1.4"

In this case, assuming `src`, `dest` and `size` are valid (and they must be, or else one may not call `ddsrt_strlcpy`), then `0 <= len && len <= srclen && len < size`. That makes all arguments valid according to the C standard.

MISRA completely ridiculously claims that 0 is not an "appropriate" value, apparently because it "may result in in read or write access beyond the bounds of an object passed as a parameter". But the the C spec says:

> On such a call, a function that locates a character finds no occurrence, a function that compares two character sequences returns zero, and a function that copies characters copies zero characters.

I've read the MISRA document in its entirety. I truly fear software written by its authors. I cannot agree to a change based on this rule where the number of bytes is 0. The most generous interpretation is that the MISRA authors don't know the difference between "positive" and "non-negative". (As if `size_t` could *ever* by negative ... but that's beside the point.)